src/ray/gcs/gcs_server/gcs_task_manager.cc

```diff
@@ -159,61 +225,76 @@ absl::optional<int64_t> GcsTaskManager::GcsTaskManagerStorage::GetTaskStatusUpda
+    const JobID &job_id, int64_t job_finish_time_ns) {
   auto task_attempts_itr = job_to_task_attempt_index_.find(job_id);
   if (task_attempts_itr == job_to_task_attempt_index_.end()) {
     // No tasks in the job.
     return;
   }
 
+  rpc::RayErrorInfo error_info;
+  error_info.set_error_type(rpc::ErrorType::WORKER_DIED);
+  std::stringstream error_message;
+  error_message << "Job finishes (" << job_id.Hex()
+                << ") as driver exits. Marking all non-terminal tasks as failed.";
+  error_info.set_error_message(error_message.str());
+
   // Iterate all task attempts from the job.
   for (const auto &task_attempt : task_attempts_itr->second) {
     if (!IsTaskTerminated(task_attempt.first)) {
-      MarkTaskAttemptFailed(task_attempt, job_finish_time_ns);
+      MarkTaskAttemptFailed(task_attempt, job_finish_time_ns, error_info);
     }
   }
 }
 
-void GcsTaskManager::GcsTaskManagerStorage::MarkTaskFailed(const TaskID &task_id,
-                                                           int64_t failed_ts) {
+void GcsTaskManager::GcsTaskManagerStorage::MarkTaskFailedOnAncestorFailed(
+    const TaskID &task_id, int64_t failed_ts, const TaskID &ancestor_task_id) {
+  RAY_LOG(DEBUG) << "Marking task " << task_id << " as failed due to ancestor task "
+                 << ancestor_task_id << " failure.";
   auto latest_task_attempt = GetLatestTaskAttempt(task_id);
   if (!latest_task_attempt.has_value()) {
     return;
   }
-  MarkTaskAttemptFailed(*latest_task_attempt, failed_ts);
+  rpc::RayErrorInfo error_info;
+  error_info.set_error_type(rpc::ErrorType::WORKER_DIED);
+  std::stringstream error_message;
+  error_message << "Ancestor task " << ancestor_task_id
+                << " failed due to unexpected worker death.";
+  error_info.set_error_type(rpc::ErrorType::WORKER_DIED);
+  error_info.set_error_message(error_message.str());
+
+  MarkTaskAttemptFailed(*latest_task_attempt, failed_ts, error_info);
 }
 
 void GcsTaskManager::GcsTaskManagerStorage::MarkTaskTreeFailedIfNeeded(
-    const TaskID &task_id, const TaskID &parent_task_id) {
-  if (!parent_task_id.IsNil()) {
-    // If parent has failed, mark itself as failed
-    auto parent_failed_ts =
-        GetTaskStatusUpdateTime(parent_task_id, rpc::TaskStatus::FAILED);
-    if (parent_failed_ts.has_value()) {
-      // Mark current task as failed.
-      MarkTaskFailed(task_id, *parent_failed_ts);
-    }
+    const TaskID &task_id) {
+  const auto &task_event = GetLatestTaskEvent(task_id);
+  if (!task_event.has_value() || !task_event->has_task_info()) {
+    return;
   }
-
-  // BFS traverse the task tree to mark all non terminal children as failure
-  std::vector<TaskID> failed_tasks;
-  auto task_failed_ts = GetTaskStatusUpdateTime(task_id, rpc::TaskStatus::FAILED);
-  if (task_failed_ts.has_value()) {
-    failed_tasks.push_back(task_id);
+  auto parent_task_id = TaskID::FromBinary(task_event->task_info().parent_task_id());
```

Comment:
Little confused about the reasoning behind these changes. Can you tell me a bit more detail?

Reply:
this is still the same BFS traversal. Before we were not able differentiate various failure conditions, and simply checking if the parent has a failed timestamp, the the traversal always starts from the child task (the `task_id`). 

Now we need both parent & child's task info in `ShouldMarkChildrenFailed` routine, thus starting from the parent task.